PMM-13154 PG17 support in QAN

artemgavrilov opened the pull request:
PMM-13154

Link to the Feature Build: SUBMODULES-3700


If this PR adds or removes or alters one or more API endpoints, please review and add or update the relevant [API documents](https://github.com/percona/pmm/tree/main/docs/api) as well:

- [ ] API Docs updated

If this PR is related to some other PRs in this or other repositories, please provide links to those PRs:

- Links to related pull requests (optional).

artemgavrilov commented on agent/runner/actions/postgresql_query_show_action_test.go:
```diff
@@ -49,13 +49,13 @@ func TestPostgreSQLQueryShow(t *testing.T) {
 		require.NoError(t, err)
 		assert.LessOrEqual(t, 22000, len(b))
-		assert.LessOrEqual(t, len(b), 33668)
+		assert.LessOrEqual(t, len(b), 37989)
```
These kind of tests seems useless for me. Not sure what it can catch.

BupycHuk replied:
agree

artemgavrilov commented on agent/agents/postgres/pgstatstatements/pgstatstatements_test.go:
```diff
@@ -408,7 +426,8 @@ func TestPGStatStatementsQAN(t *testing.T) {
 			fingerprint = fmt.Sprintf(`INSERT /* CheckMBlkReadTime controller='test' */ INTO %s (customer_id, first_name, last_name, active) VALUES ($1, $2, $3, $4)`, tableName)
 		}
 		actual := buckets[0]
-		assert.NotZero(t, actual.Postgresql.MBlkReadTimeSum)
+		assert.NotZero(t, actual.Postgresql.MSharedBlkReadTimeSum+actual.Postgresql.MSharedBlkWriteTimeSum)
+		assert.Equal(t, float32(n), actual.Postgresql.MSharedBlkReadTimeCnt+actual.Postgresql.MSharedBlkWriteTimeCnt)
```
These values are different for different versions of PG. So I'm checking combinations here to match these differences.

artemgavrilov commented on api/agentpb/collector.proto:
```diff
@@ -307,12 +307,12 @@ message MetricsBucket {
-    float m_blk_write_time_cnt = 25;
-    // Total time the statement spent writing blocks, in milliseconds (if track_io_timing is enabled, otherwise zero).
-    float m_blk_write_time_sum = 26;
+    float m_shared_blk_read_time_cnt = 23;
+    // Total time the statement spent reading shared blocks, in milliseconds (if track_io_timing is enabled, otherwise zero).
+    float m_shared_blk_read_time_sum = 24;
+    float m_shared_blk_write_time_cnt = 25;
+    // Total time the statement spent writing shared blocks, in milliseconds (if track_io_timing is enabled, otherwise zero).
+    float m_shared_blk_write_time_sum = 26;
```
It should be safe to change as agent and server don't use json API for interaction.

BupycHuk replied:
are we sure? Have you tested it?

artemgavrilov replied:
Yep, checked, it works

artemgavrilov commented:
Hey @artemgavrilov , I reworked this PR:

1. Added local_blk_read/write_time
2. Changed QAN API according to new fields, deprecated old one (but it's still working and return valid information)
3. Rewrote pg_stat_statements view reading (now it's the same as for pg_stat_monitor) to allow different schemas.

Please take a look once again

BupycHuk commented on managed/services/qan/client.go:
```diff
@@ -541,10 +541,20 @@ func fillPostgreSQL(mb *qanpb.MetricsBucket, bp *agentpb.MetricsBucket_PostgreSQ
+	mb.MLocalBlkWriteTimeSum = bp.MLocalBlkWriteTimeSum
+
+	// Fill deprecated fields for compatibility
+	mb.MBlkReadTimeCnt = bp.MSharedBlkReadTimeCnt
+	mb.MBlkReadTimeSum = bp.MSharedBlkReadTimeSum
+	mb.MBlkWriteTimeCnt = bp.MSharedBlkWriteTimeCnt
+	mb.MBlkWriteTimeSum = bp.MSharedBlkWriteTimeSum
```
I think it should work the opposite way, because we are writing to DB

artemgavrilov replied:
Discussed on call. Filling these fields are redundant, cause we keep QAN in sync with other pmm components, so it's not possible to use old QAN API and new managed.

artemgavrilov commented on managed/services/qan/client.go:
```diff
@@ -541,10 +541,20 @@ func fillPostgreSQL(mb *qanpb.MetricsBucket, bp *agentpb.MetricsBucket_PostgreSQ
+	mb.MLocalBlkWriteTimeCnt = bp.MLocalBlkWriteTimeCnt
+	mb.MLocalBlkWriteTimeSum = bp.MLocalBlkWriteTimeSum
+
+	// Fill deprecated fields for compatibility
+	mb.MBlkReadTimeCnt = bp.MSharedBlkReadTimeCnt
+	mb.MBlkReadTimeSum = bp.MSharedBlkReadTimeSum
+	mb.MBlkWriteTimeCnt = bp.MSharedBlkWriteTimeCnt
+	mb.MBlkWriteTimeSum = bp.MSharedBlkWriteTimeSum
```
```suggestion
```

artemgavrilov commented on .github/workflows/agent.yml:
```diff
@@ -26,15 +26,15 @@ jobs:
+          - { mysql: 'mysql:5.6', mongo: 'mongo:4.2', postgres: 'postgres:14', pmm_server: 'percona/pmm-server:2.0.0' }
+          - { mysql: 'mysql:5.7', mongo: 'mongo:4.4', postgres: 'postgres:15', pmm_server: 'percona/pmm-server:2.0.1' }
+          - { mysql: 'mysql:8.0', mongo: 'mongo:4.4', postgres: 'postgres:16', pmm_server: 'percona/pmm-server:2.1.0' }
+          - { mysql: 'mysql:8.0', mongo: 'mongo:4.4', postgres: 'postgres:17beta3', pmm_server: 'percona/pmm-server:2.2.0' }
```
```suggestion
          - { mysql: 'mysql:8.0', mongo: 'mongo:4.4', postgres: 'postgres:17rc1', pmm_server: 'percona/pmm-server:2.2.0' }
```
